### PR TITLE
Implement shared account pool

### PR DESCRIPTION
**Pull request by Szymongib**
**Description**

Changes proposed in this pull request:

- Implement shared account pool for HAP

**Related issue(s)**

### CODE REVIEW

File: components/kyma-environment-broker/internal/hyperscaler/shared_pool.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	if err != nil {
+		return nil,
+			errors.Wrapf(err, "sharedAccountPool error during secret list for LabelSelector: %s", labelSelector)
+	}
+	if secrets == nil || len(secrets.Items) == 0 {
+		return nil, errors.Errorf("sharedAccountPool error: no shared Secret found for %s label selector", labelSelector)
+	}
+
+	return secrets.Items, nil
+}
```
Comment (franpog859) [inline components/kyma-environment-broker/internal/hyperscaler/shared_pool.go:94]:
```suggestion
func getK8sSecrets(secretsClient corev1.SecretInterface, labelSelector string) ([]apiv1.Secret, error) {
	secrets, err := secretsClient.List(metav1.ListOptions{
		LabelSelector: labelSelector,
	})
	if err != nil {
		return nil,
			errors.Wrapf(err, "sharedAccountPool error during secret list for LabelSelector: %s", labelSelector)
	}
	if secrets == nil || len(secrets.Items) == 0 {
		return nil, errors.Errorf("sharedAccountPool error: no shared Secret found for %s label selector", labelSelector)
	}

	return secrets.Items, nil
}

func (sp *SharedAccountPool) getLeastUsed(secrets []apiv1.Secret) (apiv1.Secret, error) {
	usageCount := make(map[string]int, len(secrets))
	for _, s := range secrets {
		usageCount[s.Name] = 0
	}

	shoots, err := sp.shootsClient.List(metav1.ListOptions{})
	if err != nil {
		return apiv1.Secret{}, errors.Wrap(err, "sharedAccountPool error while listing Shoots")
	}

	if shoots == nil || len(shoots.Items) == 0 {
		return secrets[0], nil
	}

	for _, s := range shoots.Items {
		count, found := usageCount[s.Spec.SecretBindingName]
		if !found {
			continue
		}

		usageCount[s.Spec.SecretBindingName] = count + 1
	}

	min := usageCount[secrets[0].Name]
	minIndex := 0

	for i, s := range secrets {
		if usageCount[s.Name] < min {
			min = usageCount[s.Name]
			minIndex = i
		}
	}

	return secrets[minIndex], nil
}
```

File: components/kyma-environment-broker/internal/hyperscaler/shared_pool.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+)
+
+type SharedPool interface {
+	SharedCredentials(hyperscalerType Type) (Credentials, error)
+}
+
+func NewSharedGardenerAccountPool(secretsClient corev1.SecretInterface, shootClient gardener_apis.ShootInterface) *SharedAccountPool {
+	return &SharedAccountPool{
+		secretsClient: secretsClient,
+		shootsClient:  shootClient,
```
Comment (franpog859) [inline components/kyma-environment-broker/internal/hyperscaler/shared_pool.go:20]:
```suggestion
func NewSharedGardenerAccountPool(secretsClient corev1.SecretInterface, shootsClient gardener_apis.ShootInterface) *SharedAccountPool {
	return &SharedAccountPool{
		secretsClient: secretsClient,
		shootsClient:  shootsClient,
```

File: components/kyma-environment-broker/internal/broker/plans.go [hunk 3]
```diff
@@ -9,6 +9,8 @@ import (
 const (
 	GCPPlanID         = "ca6e5357-707f-4565-bbbd-b3ab732597c6"
 	GCPPlanName       = "gcp"
+	GCPTrialPlanID    = "7d55d31d-35ae-4438-bf13-6ffdfa107d9f"
+	GCPTrialPlanName  = "gcp_trial"
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/broker/plans.go:13]: the name should be "trial"

File: components/kyma-environment-broker/internal/gardener/client.go [hunk 4]
```diff
@@ -38,6 +38,18 @@ func NewGardenerSecretsInterface(gardenerClusterCfg *restclient.Config, gardener
 	return gardenerClusterClient.CoreV1().Secrets(gardenerNamespace), nil
 }
 
+func NewGardenerShootInterface(gardenerClusterCfg *restclient.Config, gardenerProjectName string) (gardener_apis.ShootInterface, error) {
+
+	gardenerNamespace := fmt.Sprintf("garden-%s", gardenerProjectName)
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/gardener/client.go:43]: minor: there are two places which creates namespace using the pattern, please extract it to a separate function

File: components/kyma-environment-broker/internal/broker/plans.go (hunk 3 above)
Comment (Szymongib) [reply components/kyma-environment-broker/internal/broker/plans.go:13]: Wouldn't it be better to have the distinction between hyperscalers? I guess we might want to introduce a trial on different providers in the future.

**Pull request merged**